### PR TITLE
MINOR: Test SASL authorization id

### PR DESCRIPTION
**Pull request by rajinisivaram**

### CODE REVIEW

**Comment by rajinisivaram** [comment]
@ijuma Can you review, please? Thank you!

File: clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java [hunk 1]
```diff
@@ -79,7 +79,7 @@ public PlainSaslServer(JaasContext jaasContext) {
         }
         if (tokens.length != 3)
             throw new SaslException("Invalid SASL/PLAIN response: expected 3 tokens, got " + tokens.length);
-        authorizationID = tokens[0];
+        String authorizationIDFromClient = tokens[0];
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java]: Nit: should probably `authorizationIdFromClient` to follow our convention.

File: clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java [hunk 2]
```diff
@@ -89,14 +89,18 @@ public PlainSaslServer(JaasContext jaasContext) {
...
         String expectedPassword = jaasContext.configEntryOption(JAAS_USER_PREFIX + username,
                 PlainLoginModule.class.getName());
         if (!password.equals(expectedPassword)) {
             throw new SaslException("Authentication failed: Invalid username or password");
         }
+
+        if (!authorizationIDFromClient.isEmpty() && !authorizationIDFromClient.equals(username))
+            throw new SaslException("Authentication failed: Client requested an authorization id that is different from username");
+
+        this.authorizationID = username;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java]: This should probably be `authorizationId` to follow our convention.

File: clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java [hunk 3]
```diff
@@ -90,6 +90,10 @@ public ScramSaslServer(ScramMechanism mechanism, Map<String, ?> props, CallbackH
                         this.scramCredential = credentialCallback.scramCredential();
                         if (scramCredential == null)
                             throw new SaslException("Authentication failed: Invalid user credentials");
+                        String authorizationIDFromClient = clientFirstMessage.authorizationId();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java]: A couple of questions:
1. Do we need to store `serverNonce` as a field?
2. Should we clear `clientFirstMessage` and `serverFirstMessage` after they're used?

File: clients/src/test/java/org/apache/kafka/common/security/plain/PlainSaslServerTest.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        options.put("user_" + USER_A, PASSWORD_A);
+        options.put("user_" + USER_B, PASSWORD_B);
+        jaasConfig.addEntry("jaasContext", PlainLoginModule.class.getName(), options);
+        JaasContext jaasContext = new JaasContext("jaasContext", JaasContext.Type.SERVER, jaasConfig);
+        saslServer = new PlainSaslServer(jaasContext);
+    }
+
+    @Test
+    public void noAuthorizationIdSpecified() throws Exception {
+        saslServer.evaluateResponse(saslMessage("", USER_A, PASSWORD_A));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/security/plain/PlainSaslServerTest.java]: Should we assert something on the result of the call? Same for the other cases where no exception is thrown.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, just a few minor comments.

File: clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java (hunk 1 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java]: Done.

File: clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java (hunk 2 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/security/plain/PlainSaslServer.java]: Done.

File: clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java (hunk 3 above)
Comment (rajinisivaram) [reply clients/src/main/java/org/apache/kafka/common/security/scram/ScramSaslServer.java]: Made `serverNonce` a local variable, reset `scramCredentials`, `clientFirstMessage` and `serverFirstMessage` after use.

File: clients/src/test/java/org/apache/kafka/common/security/plain/PlainSaslServerTest.java (hunk 4 above)
Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/security/plain/PlainSaslServerTest.java]: Done.

**Comment by rajinisivaram** [comment]
@ijuma Thank you for the review. Have made the updates.

File: clients/src/test/java/org/apache/kafka/common/security/scram/ScramSaslServerTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        String nonce = formatter.secureRandomString();
+        String firstMessage = String.format("n,,n=%s,r=%s", USER_A, nonce);
+        byte[] nextChallenge = saslServer.evaluateResponse(firstMessage.getBytes(StandardCharsets.UTF_8));
+        assertTrue("Next challenge is empty", nextChallenge.length > 0);
+    }
+
+    @Test
+    public void authorizatonIdEqualsAuthenticationId() throws Exception {
+        String nonce = formatter.secureRandomString();
+        String firstMessage = String.format("n,a=%s,n=%s,r=%s", USER_A, USER_A, nonce);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/security/scram/ScramSaslServerTest.java]: Nit: should we have a helper for generating the message like you did in the other test?

Comment (rajinisivaram) [reply clients/src/test/java/org/apache/kafka/common/security/scram/ScramSaslServerTest.java]: Done, thanks.

**Review by ijuma — APPROVED** [review]
LGTM, thanks.

**Pull request closed without merging**